ffi_internals/src/impl_internals/impl_ffi.rs

```diff
@@ -0,0 +1,206 @@
+                        syn::ReturnType::Type(_token, ty) => Some(NativeTypeData::from(
+                            UnparsedNativeTypeData::initial(*ty.clone()),
+                        )),
+                    };
+
+                    Some(FnFFI {
+                        fn_name,
+                        has_receiver,
+                        parameters: arguments,
+                        return_type,
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Self {
+            trait_name: inputs.trait_name,
+            type_name: inputs.type_name,
+            fns,
+            import_paths: inputs.import_paths,
+        }
+    }
+}
+
+/// A representation of a Rust impl that can be used to generate an FFI and consumer code for
+/// calling that FFI.
+///
+#[derive(Debug)]
+pub struct ImplFFI {
+    /// The name of the trait that's implemented.
+    ///
+    /// Note that this is currently required; we don't support standalone impls right now because
+    /// we're relying on the trait name + type name pair to guarantee uniqueness for the generated
+    /// FFI module and consumer file. If we have a use case for exposing standalone impls, we'll
+    /// have to come up with another way to ensure that uniqueness.
+    ///
+    trait_name: Ident,
+
+    /// The name of the type that this implementation applies to.
+    ///
+    type_name: Ident,
+
+    /// A collection of representations of the functions declared in this impl that can be used to
+    /// generate an FFI and consumer code for each function.
+    ///
+    fns: Vec<FnFFI>,
+
+    /// Any import paths specified in the attributes on the macro invocation.
+    ///
+    import_paths: Vec<Path>,
+}
+
+impl ImplFFI {
+    pub fn consumer_file_name(&self) -> String {
+        format!("{}_{}.swift", self.trait_name, self.type_name)
+    }
+
+    /// Generates an implementation for the consumer's type so that they'll be able to call it like
+    /// `nativeTypeInstance.someMethod(with: params)`. Hardcoded to Swift for now like all the other
+    /// consumer output, until we bother templating for other languages.
+    ///
+    /// Example output:
+    /// ```ignore
+    /// extension SelectedField {
+    ///     func build_commodity_locations(plantings: [CLPlanting]) -> [CommodityLocation] {
+    ///         [CommodityLocation].fromRust(build_commodity_locations(pointer, plantings.toRust()))
+    ///     }
+    /// }
+    /// ```
+    ///
+    pub fn generate_consumer(&self) -> String {
+        format!(
+            r#"
+extension {native_type} {{
+    {functions}
+}}
+            "#,
+            native_type = self.type_name.to_string(),
+            functions = self
+                .fns
+                .iter()
+                .map(|f| f.generate_consumer())
+                .collect::<Vec<String>>()
+                .join("\n"),
+        )
+    }
```

Comment:
Surely it's non-trival, but my crazy brain just had me wondering if there'd be good ROI on making a `quote!()` that works with Swift code instead of Rust code? Or something along the lines. I suppose that'd entail writing both a parser and a token generator... meh, maybe not.

Reply:
That seems like it'd be super nice to have...lots of work to write, but it would make these easier to write and maintain (especially compared to doing these `format!` templates with multiple languages if we get there...)